src/pybind/mgr/mirroring/fs/snapshot_mirror.py

```diff
@@ -0,0 +1,579 @@
+        except Exception as e:
+            return e.args[0], '', 'failed to disable mirroring'
+
+    def peer_add(self, filesystem, remote_cluster_spec, remote_fs_name):
+        try:
+            if remote_fs_name == None:
+                remote_fs_name = filesystem
+            with self.lock:
+                fspolicy = self.pool_policy.get(filesystem, None)
+                if not fspolicy:
+                    raise MirrorException(-errno.EINVAL, f'filesystem {filesystem} is not mirrored')
+                self.verify_and_set_mirror_info(filesystem, remote_cluster_spec, remote_fs_name)
+                cmd = {'prefix': 'fs mirror peer_add',
+                       'fs_name': filesystem,
+                       'remote_cluster_spec': remote_cluster_spec,
+                       'remote_fs_name': remote_fs_name}
+                r, outs, err = self.mgr.mon_command(cmd)
+                if r < 0:
+                    log.error(f'mon command to add peer failed: {err}')
+                    raise Exception(-errno.EINVAL)
+                return 0, json.dumps({}), ''
+        except MirrorException as me:
+            return me.args[0], '', me.args[1]
+        except Exception as e:
+            return e.args[0], '', 'failed to add peer'
+
+    def peer_remove(self, filesystem, peer_uuid):
+        try:
+            with self.lock:
+                fspolicy = self.pool_policy.get(filesystem, None)
+                if not fspolicy:
+                    raise MirrorException(-errno.EINVAL, f'filesystem {filesystem} is not mirrored')
+                cmd = {'prefix': 'fs mirror peer_remove',
+                       'fs_name': filesystem,
+                       'uuid': peer_uuid}
+                r, outs, err = self.mgr.mon_command(cmd)
+                if r < 0:
+                    log.error(f'mon command to remove peer failed: {err}')
+                    raise Exception(-errno.EINVAL)
+                return 0, json.dumps({}), ''
+        except MirrorException as me:
+            return me.args[0], '', me.args[1]
+        except Exception as e:
+            return e.args[0], '', 'failed to remove peer'
+
+    def verify_and_normalize_dir(self, fs_handle, dir_path):
+        norm_path = os.path.normpath(dir_path)
+        try:
+            stx = fs_handle.statx(norm_path, cephfs.CEPH_STATX_MODE, cephfs.AT_SYMLINK_NOFOLLOW)
+            if not stat.S_ISDIR(stx["mode"]):
+                raise MirrorException(-errno.EINVAL, f'path {dir_path} should be a directory')
+        except cephfs.ObjectNotFound:
+            raise MirrorException(-errno.ENOENT, f'{dir_path} does not exist')
+        except cephfs.Error as e:
+            log.error(f'failed to verify directory {dir_path}: {e}')
+            raise Exception(-e.errno)
+        return norm_path
+
+    def verify_and_normalize_path(self, filesystem, dir_path):
+        if not os.path.isabs(dir_path):
+            raise MirrorException(-errno.EINVAL, f'{dir_path} should be an absolute path')
```

Comment:
nit: you probably want to quote the names of the filesystems and directory paths where ever they are printed

Reply:
sure